Serialisation: add a built-in serialisation provider.

paulidale opened the pull request:
Move the libcrypto serialisation functionality into a place where it can be provided at some point.  The serialisation still remains native in the default provider.

Add additional code to the _list_ command to display what kind of serialisation each entry is capable of.

Having the FIPS provider auto load the serialisation provider is a future (but necessary) enhancement.

- [x] documentation is added or updated
- [ ] tests are added or updated

paulidale commented:
This separates out the serialisation into a separate provider, however it lives in libcrypto and is included in the default provider automatically.  The FIPS provider needs to load this when it is loaded and unload it when it is unloaded.

Added a `-serialisers` option to the _list_ app.  This includes extra information about the format and type of serialisation supported.

slontis commented on providers/build.info:
```diff
@@ -88,6 +88,16 @@ INCLUDE[$DEFAULTGOAL]=implementations/include
+#
+# Serialisation provider stuff
+#
+# Because the derialisation provider is build it, it means that libcrypto must
```
der?

paulidale replied:
It should have been and now is: `deser`

levitte commented on apps/list.c:
```diff
@@ -277,6 +279,73 @@ static void list_kdfs(void)
+        print_names(bio_out, names);
+        BIO_printf(bio_out, " @ %s for doing %s to %s serialisation\n",
+                   OSSL_PROVIDER_name(OSSL_SERIALIZER_provider(k)),
+                   type, format);
```
I disagree with this output.  It would be much simpler, *and* more useful for programmers if you simply printed what `OSSL_SERIALIZER_properties()` returns.

paulidale replied:
This simplified the code a lot.  Not quite so nice for the user.

levitte commented on providers/defltprov.c:
```diff
@@ -374,154 +374,6 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
-#endif
-    { NULL, NULL, NULL }
-};
-
```
I hope that everyone understands that with this removal, *every* serialization (and deserialization) will imply a keydata export (from keymgmt)/import (into serializer).

Churn, churn, churn...

paulidale replied:
The default provider still includes the serialisers -- there will be no unnecessary import/export for it.

levitte replied:
Huh?????

...

Ohhhh!  I didn't notice the change in `deflt_query`!

paulidale replied:
Neat huh?

paulidale commented:
Note: there appears to be a lock order inversion that is trigger by this change.

paulidale commented:
The threading issue is reported in #12151 (which is on master).

paulidale commented:
Out of WIP.

The threading issue that caused the lock up was a result of a problem with this PR (i.e. the issue existed regardless but was triggered due to a problem here).

The "provider=" properties for the serialisers is fixed as _serialization_ even if it comes from the default provider.  I don't see this as being a problem, but if others disagree, it can be fixed easily.

levitte commented:
> The "provider=" properties for the serialisers is fixed as _serialization_ even if it comes from the default provider.  I don't see this as being a problem, but if others disagree, it can be fixed easily.

That has me wonder what the meaning is with that particular property.  Why have it at all?

levitte commented on doc/man1/openssl-list.pod.in:
```diff
@@ -73,6 +74,22 @@ displayed.
+In both cases, C<bar> is the name of the provider and C<properties> are the
+property settings for the serializer.  The C<type> and C<format> properties
+distinguish between otherwise identical serializers.
+
```
Why do you repeat here what's explained *generically* in "Display of algorithm names"?  The following should be enough here:

```
=item B<-serializers>

Display a list of serializers.
See L</Display of algorithm names> for a description of how names are
displayed.

In verbose mode, the algorithms provided by a provider will get additional
information on what parameters each implementation supports.
```
```

paulidale replied:
Changed.

levitte commented on doc/man7/OSSL_PROVIDER-default.pod:
```diff
@@ -192,7 +192,9 @@ The OpenSSL default provider supports these operations and algorithms:
-In addition to "provider=default", this set of implementations define the
+The default provider also includes all of the serialization algorithms
+present in the serialization provider.  These are defined with the
+"provider=serialization" property and some additionally have the
```
Yeah, no, this doesn't sit well with me

paulidale replied:
Okay, will fix.

paulidale commented:
We've got "provider=XXX" for all other instances.  In this case the provider is the _serialization_ provider even if it isn't loaded.

As I mentioned, I'm happy to change this to _provider=default_.

Regardless, _provider=serialization_ will always be exactly equivalent to _provider=default_.  Examine the output of `openssl list --serializations` both with and without `-provider serialization` specified.

levitte commented:
I noticed, with some amusement, the spelling change from "serialise" to "serialize"...  I actually thought the 's' spelling was favoured in AU :wink:
But ah well, the functionality is spelt with a 'z', so we might as well adopt that everywhere, even though we usually have a slight lean toward current British spelling...

For further amusement: https://english.stackexchange.com/questions/38232/usage-of-z-in-the-word-serialized-in-english

paulidale commented:
English uses 'z' and 's' in -[sz]se pretty much interchangeably, alhtough 's' is generally preferred.  American only uses 'z'.  My preference is towards 's'.

However, there were plenty of instances using 'z' already in the code, which made standardising on that reasonable.  Yes, I used 's' deliberately :)

paulidale commented:
Ping @openssl/committers

kaduk commented:
> However, there were plenty of instances using 'z' already in the code, which made standardising on that reasonable. Yes, I used 's' deliberately :)

Reminds me of how yesterday I got a build error in my new code by putting in a check against `IV_STATE_UNINITIALIZED` when the actual macro uses the 'S' spelling.  Perhaps, in the case of macros, we should provide both spellings?

paulidale commented:
> That has me wonder what the meaning is with that particular property. Why have it at all?

I'm happy to duplicate the table, I don't really see a need for this property.  People want an implementation that conforms to X, the provider should be irrelevant.  It was introduced so that the default and FIPS providers could coexist and share the serialisation.  I'd vote to remove it -- albeit not as part of this PR.

paulidale commented:
> I noticed, with some amusement, the spelling change from "serialise" to "serialize"... I actually thought the 's' spelling was favoured in AU 😉

In English, either `z` or `s` are allowed in the `-i[zs]e` word ending.  For the most part the `s` is preferred but both are permissible (with a few exceptions of course, this is English).  Australian follows this convention as well but with an ever so slightly stronger influence for the American `-ize`.

I went through the code and noticed that there were far more `serialize` (typo checking changed that to the `-ise` form if that's any indication) than `serialise`, so I changed things to the former.  My opinion is towards the `-ise` version but not strongly enough to cause a fuss.

That written, I'll happily change all to the `-ise` form if we want.  I think her Majesty's English is what we prefer.  Update on Monday.

levitte commented:
> the provider should be irrelevant.

Yes, but if we do specify it, it should be correct.

levitte commented:
> I think her Majesty's English is what we prefer.

Most of us at least :wink:

paulidale commented:
Review comments addressed.

Apart from unifying the spelling of all the -ise and -ize words.  I'm happy to drop the changes I did in this area so that something with wider coverage can be done in the future.

paulidale commented:
Ping for review.

slontis commented on apps/list.c:
```diff
@@ -277,6 +279,63 @@ static void list_kdfs(void)
+    STACK_OF(OSSL_SERIALIZER) *serializers;
+    int i;
+
+    serializers = sk_OSSL_SERIALIZER_new(serializer_cmp);
```
add a null check here..

paulidale replied:
Added NULL checks for all the cases above too.  There were none.

slontis commented on apps/list.c:
```diff
@@ -619,7 +678,7 @@ typedef enum HELPLIST_CHOICE {
     OPT_DIGEST_ALGORITHMS, OPT_CIPHER_COMMANDS, OPT_CIPHER_ALGORITHMS,
     OPT_PK_ALGORITHMS, OPT_PK_METHOD, OPT_ENGINES, OPT_DISABLED,
-    OPT_KDF_ALGORITHMS, OPT_MISSING_HELP, OPT_OBJECTS,
+    OPT_KDF_ALGORITHMS, OPT_SERIALIZERS, OPT_MISSING_HELP, OPT_OBJECTS,
```
please add a test..

paulidale replied:
I'm not sure what to test.  None of the other `list` options seems to be tested.

slontis commented on doc/man7/OSSL_PROVIDER-serialization.pod:
```diff
@@ -0,0 +1,99 @@
+=back
+
+These may be used in a property query string with fetching functions to select
+the serialization output format.  Either the DER, PEM and plain test are
```
plaintext

slontis commented on doc/man7/OSSL_PROVIDER-serialization.pod:
```diff
@@ -0,0 +1,99 @@
+
+L<OSSL_PROVIDER-default(7)>, L<openssl-core.h(7)>,
+L<openssl-core_numbers.h(7)>, L<provider(7)>
+
```
HIstory section?

paulidale replied:
There are no history sections in the other OSSL_PROVIDER PODs.

slontis commented on doc/man7/provider.pod:
```diff
@@ -260,6 +260,14 @@ algorithm identifier to the appropriate fetching function.
+
+The serialisation provider is built in as part of the F<libcrypto> library.
+Should it be needed (if other providers are loaded and offer
+implementations of the same algorithms), the property "provider=serialisation"
```
so now you spell it here with a s :)

slontis commented on providers/defltprov.c:
```diff
@@ -374,153 +374,16 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
-#endif
+#define SER(name, fips, format, type, func_table)                           \
+    { name,                                                                 \
+      "provider=default,fips=" fips ",format=" format ",type=" type,  \
```
These " look funny here, is this what you intended?

paulidale replied:
Yes, they provide a visual break between the macro argument and the string.

slontis commented on providers/defltprov.c:
```diff
@@ -374,153 +374,16 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
-    { "EC", "provider=default,fips=yes,format=pem,type=parameters",
-      ec_param_pem_serializer_functions },
-#endif
+#include "serialisers.h"
```
hehe - the s spelling again :)

paulidale replied:
It's what I naturally type :(

paulidale replied:
Changed.

paulidale commented:
Conflicts resolved and rebased to master.  Ping for review?
Do we want this in alpha 4?

paulidale commented:
Ping for review

paulidale commented:
Reping.

paulidale commented:
Ping @openssl/committers for review.

paulidale commented:
0 bytes from @openssl/committers : icmp_seq=3 ttl=64 time=172800000 ms

mattcaswell commented on apps/list.c:
```diff
@@ -70,6 +72,10 @@ static void list_ciphers(void)
+    if (ciphers == NULL) {
+        BIO_printf(bio_err, "ERROR: Memory allocation\n");
+        return;
+    }
```
The seems to be a repetition of the code just a few lines above?

paulidale replied:
A merge addition.  Will remove.

mattcaswell commented on apps/list.c:
```diff
@@ -145,6 +151,10 @@ static void list_digests(void)
+    if (digests == NULL) {
+        BIO_printf(bio_err, "ERROR: Memory allocation\n");
+        return;
+    }
```
Same here?

mattcaswell commented on providers/build.info:
```diff
@@ -88,6 +88,16 @@ INCLUDE[$DEFAULTGOAL]=implementations/include
+#
+# Serialization provider stuff
+#
+# Because the serialization provider is build in, it means that libcrypto
```
Typo: built in

paulidale replied:
Fixed.

paulidale commented:
Comments addressed.

paulidale commented:
@levitte I've renamed the provider to _base_ as discussed.

mattcaswell commented on doc/man7/provider.pod:
```diff
@@ -268,6 +268,14 @@ algorithm identifier to the appropriate fetching function.
+
+The serialization provider is built in as part of the F<libcrypto> library.
+Should it be needed (if other providers are loaded and offer
+implementations of the same algorithms), the property "provider=serialization"
```
Drive by comment: base?

paulidale replied:
Yeah, should have been "base" in several places in this paragraph.  All fixed now.

levitte commented on crypto/provider_predefined.c:
```diff
@@ -21,6 +22,7 @@ const struct predefined_providers_st predefined_providers[] = {
     { "fips", fips_intern_provider_init, 1 },
 #else
     { "default", ossl_default_provider_init, 1 },
+    { "base", ossl_base_provider_init, 0 },
```
Hmmm, this means that it's only pre-defined, but not activated...  on the other hand, setting the flag to 1 only has it auto-activated if no provider has been explicitly loaded beforehand.

It's not strictly necessary to resolve that in *this* PR, and I have an idea on how to do it, if we can have the patience to wait (I don't even have a branch yet, for once)

paulidale replied:
The default provider doesn't need this and it shouldn't be automatically loaded with the default for performance reasons.  I think it should load with the FIPS provider automatically, but that's for another PR.

paulidale replied:
Actually, the base is in the wrong place...  will move.

paulidale commented:
Travis isn't relevant.

paulidale commented:
Addressed conflicts and rebased to master.  Ping for review?

paulidale commented:
Re-ping?

paulidale commented:
Ping @openssl/committers for review

slontis commented:
Needs a rebase..
#12162 depends on the serializer being seperate (i.e so fips provider can work inpendantly of default provider).

slontis commented on providers/serializers.h:
```diff
@@ -0,0 +1,102 @@
+/*
```
I prefer calling these .inc files when they are just going to be embedded in another file.

paulidale replied:
Renamed.

slontis commented:
I assume deserialize will be added in a separate PR?

paulidale commented:
> I assume deserialize will be added in a separate PR?

Presumably.  They could be added here if necessary.

paulidale commented:
Conflict addressed and rebased to master.  Still awaiting review.

mattcaswell commented on doc/man7/OSSL_PROVIDER-default.pod:
```diff
@@ -192,7 +192,8 @@ The OpenSSL default provider supports these operations and algorithms:
 
-In addition to "provider=default", this set of implementations define the
+The default provider also includes all of the serialization algorithms
+present in the serialization provider.  Some of these have the
```
base provider?

paulidale replied:
Fixed.

mattcaswell commented on doc/man7/OSSL_PROVIDER-serialization.pod:
```diff
@@ -0,0 +1,99 @@
+
+=head1 NAME
+
+OSSL_PROVIDER-serialization - OpenSSL serialization provider
```
Base?...and so on in this file (and the file name is wrong too)

paulidale replied:
Fixed the file name.

mattcaswell commented on doc/man7/OSSL_PROVIDER-serialization.pod:
```diff
@@ -0,0 +1,99 @@
+implementations of this provider and none other.
+
+The serialization algorithms in the default provider also have
+provider=serialization set.
```
Hmmmm. Really? This seems wrong (and doesn't seem to match what is implemented).

paulidale replied:
Oops....

mattcaswell commented on doc/man7/OSSL_PROVIDER-serialization.pod:
```diff
@@ -0,0 +1,99 @@
+
+=head1 OPERATIONS AND ALGORITHMS
+
+The OpenSSL default provider supports these operations and algorithms:
```
Now we're talking about the default provider

mattcaswell commented on doc/man7/OSSL_PROVIDER-serialization.pod:
```diff
@@ -0,0 +1,99 @@
+=head2 Asymmetric Key Serializer
+
+In addition to "provider=serialization", some of these serializers define the
+property "fips=yes", to allow them to be used together with the FIPS
```
Some? Which ones?

paulidale replied:
Is a list required?  The default provider says the same thing without giving a list.

mattcaswell replied:
Ok

mattcaswell commented on doc/man7/provider.pod:
```diff
@@ -268,6 +268,14 @@ algorithm identifier to the appropriate fetching function.
 Should it be needed (if other providers are loaded and offer
 implementations of the same algorithms), the property "provider=default"
+can be used as a search criterion for these implementations.  The default
+provider includes all the functionality of the serialization provider below.
```
base

paulidale commented:
All bar listing the FIPS serialisers fixed.

mattcaswell commented:
doc-nits failure reported by travis sees relevant. The test_verify failure OTOH looks odd. Several failures due to that but I can't see how it is related.

mattcaswell commented:
Looks good except for the travis issues.

mattcaswell commented:
> The test_verify failure OTOH looks odd.

This is fixed by #12549

paulidale commented:
Had to miss something.  Should be fixed now.

paulidale commented:
I've squashed an rebased the changes -- the build failure was a link fix in the documentation.

I've also added a fixup commit that adds a `-deserializers` option to the _list_ command and this will require review.

paulidale commented:
Travis showed up twice when I last pushed.  Both builds have the same ID (36399).

I've no idea why this happened.

slontis commented:
> I've no idea why this happened.

https://travis-ci.community/t/build-config-running-twice/6985

paulidale commented:
Merged to master.  Thanks for all the reviews etc.
